relay-server/src/utils/metrics_rate_limits.rs

```diff
@@ -0,0 +1,166 @@
+                    // so we can use it to count the number of transactions.
+                    let count = bucket.value.len();
+                    Some(count)
+                } else {
+                    // For any other metric in the transaction namespace, we check the limit with quantity=0
+                    // so transactions are not double counted against the quota.
+                    Some(0)
+                }
+            })
+            .collect();
+
+        // Accumulate the total transaction count:
+        let transaction_count = transaction_counts
+            .iter()
+            .fold(None, |acc, transaction_count| match transaction_count {
+                Some(count) => Some(acc.unwrap_or(0) + count),
+                None => acc,
+            });
+
+        if let Some(transaction_count) = transaction_count {
+            let transaction_buckets = transaction_counts.iter().map(Option::is_some).collect();
+            Ok(Self {
+                buckets,
+                quotas,
+                scoping,
+                transaction_buckets,
+                transaction_count,
+            })
+        } else {
+            Err(buckets)
+        }
+    }
+
+    #[allow(dead_code)]
+    pub fn scoping(&self) -> &Scoping {
+        &self.scoping
+    }
+
+    #[allow(dead_code)]
+    pub fn quotas(&self) -> &[Quota] {
+        self.quotas.as_ref()
+    }
+
+    #[allow(dead_code)]
+    pub fn transaction_count(&self) -> usize {
+        self.transaction_count
+    }
+
+    fn drop_with_outcome(&mut self, outcome: Outcome) {
+        // Drop transaction buckets:
+        let buckets = std::mem::take(&mut self.buckets);
+
+        self.buckets = buckets
+            .into_iter()
+            .zip(self.transaction_buckets.iter())
+            .filter_map(|(bucket, is_transaction_bucket)| {
+                (!is_transaction_bucket).then_some(bucket)
+            })
+            .collect();
+
+        // Track outcome for the processed transactions we dropped here:
+        if self.transaction_count > 0 {
+            TrackOutcome::from_registry().send(TrackOutcome {
+                timestamp: UnixTimestamp::now().as_datetime(), // as good as any timestamp
+                scoping: self.scoping,
+                outcome,
+                event_id: None,
+                remote_addr: None,
+                category: DataCategory::TransactionProcessed,
+                quantity: self.transaction_count as u32,
+            });
+        }
+    }
+
+    // Drop transaction-related buckets and create outcomes for any active rate limits.
+    //
+    // If rate limits could not be checked for some reason, pass an `Err` to this function.
+    // In this case, transaction-related metrics buckets will also be dropped, and an "internal"
+    // outcome is generated.
+    //
+    // Returns true if any buckets were dropped.
+    pub fn enforce_limits(&mut self, rate_limits: Result<&RateLimits, ()>) -> bool {
```

Comment:
To accept the result into the function seems a little bit strange and we kind of ignore the `Err` anyway.
Could we just accept `Option<&RateLimits>` instead? In my opinion it will make the interface a little bit cleaner.

Reply:
I see your point, given that the `Err` variant is empty anyway. But I feel that using a `Result` instead of an `Option` makes it clear to the reader that this is an error case.

Reply:
When I look into this, first comes to my mind that all the errors should have been handled before calling this function, and it should get here only if we are ok with enforcing the limits (whatever it means in this case).  

If we have `Option`, in case of `None` it says to me - there is no limits to enforce, just emit outcome, otherwise enforce what you can 😄 once there is `Some(_)`